[AutoDiff] Simplify varied propagation in activity analysis.

dan-zheng opened the pull request:
- Remove `setVaried`.
  - `setVaried` has only one user and can be inlined. Addresses [feedback](https://github.com/apple/swift/pull/28155#discussion_r344401285) from @marcrasi.
- Remove `setVariedAcrossArrayInitialization`.
  - Activity info does not change when removing special variedness propagation
    support for `array.uninitialized_intrinsic` applications.
  - Verified with array initialization activity info test case.

dan-zheng committed:
[AutoDiff] Simplify varied propagation in activity analysis.

- Remove `setVaried`.
  - `setVaried` has only one user and can be inlined.
- Remove `setVariedAcrossArrayInitialization`.
  - Activity info does not change when removing special variedness propagation
    support for `array.uninitialized_intrinsic` applications.
  - Verified with array initialization activity info test case.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
> Activity info does not change when removing special variedness propagation
support for array.uninitialized_intrinsic applications.
> Verified with array initialization activity info test case.

Could you please add a test to make sure this still holds true for indirectly initialized array literals? Indirectness should always be considered in these sorts of test cases.

```swift
func testArrayUninitializedIntrinsic<T>(_ x: T, _ y: T) -> [T] {
   return [x, y]
}
```

dan-zheng committed:
Add activity info test for indirect `array.uninitialized_intrinsic` application.

Remove unused comments.

dan-zheng commented:
> Could you please add a test to make sure this still holds true for indirectly initialized array literals? Indirectness should always be considered in these sorts of test cases.

Done in https://github.com/apple/swift/pull/28191/commits/05122fbe7fa5f0fec68f67f60f660a9427b1b0ae. Verified that activity info for `array.uninitialized_intrinsic` applications didn't change before/after this patch.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
Great. To throw another example here: Consider testing an indirect type (e.g. `struct S<T> { var x: T }`) that conforms to `ExpressibleByArrayLiteral`.

dan-zheng committed:
Add indirect `ExpressibleByArrayLiteral` test.

No activity analysis test added because activity info for the
`array.uninitialized_intrinsic` application is the same.

dan-zheng commented:
> Great. To throw another example here: Consider testing an indirect type (e.g. `struct S<T> { var x: T }`) that conforms to `ExpressibleByArrayLiteral`. This is different from `Array` because `Array` is always directly passed.

Added an "indirect `ExpressibleByArrayLiteral` type" test in https://github.com/apple/swift/pull/28191/commits/bb572417e07586560272e193009cad92e518b633.

I didn't add a corresponding activity analysis test because activity info around the `array.uninitialized_intrinsic` application is the same.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
